File: app/src/main/java/com/redteamobile/smart/cellular/TelephonySettingImpl.java
package com.redteamobile.smart.cellular;

import android.annotation.SuppressLint;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.telephony.SubscriptionInfo;
import android.telephony.SubscriptionManager;
import android.telephony.TelephonyManager;
import android.text.TextUtils;
import android.util.Log;

import com.redteamobile.smart.util.LogUtil;
import com.redteamobile.smart.util.SharePrefSetting;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import static android.telephony.TelephonyManager.NETWORK_TYPE_1xRTT;
import static android.telephony.TelephonyManager.NETWORK_TYPE_CDMA;
import static android.telephony.TelephonyManager.NETWORK_TYPE_EDGE;
import static android.telephony.TelephonyManager.NETWORK_TYPE_EVDO_A;
import static android.telephony.TelephonyManager.NETWORK_TYPE_GPRS;
import static android.telephony.TelephonyManager.NETWORK_TYPE_GSM;
import static android.telephony.TelephonyManager.NETWORK_TYPE_IDEN;
import static android.telephony.TelephonyManager.NETWORK_TYPE_IWLAN;
import static android.telephony.TelephonyManager.NETWORK_TYPE_LTE;
import static android.telephony.TelephonyManager.NETWORK_TYPE_TD_SCDMA;

public class TelephonySettingImpl implements TelephonySetting {

    private static final String TAG = "TelephonySettingImpl";
    private static final String APN_NAME = "Redtea Mobile";
    private static final Uri APN_URI = Uri.parse("content://telephony/carriers");
    private static final Uri PREFERRED_APN_URI = Uri.parse("content://telephony/carriers/preferapn");

    private TelephonyManager telephonyManager;
    private SubscriptionManager subscriptionManager;
    private ConnectivityManager connectivityManager;
    private ContentResolver resolver;
    private boolean isMultiSim;

    public TelephonySettingImpl(Context context) {
        telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
        connectivityManager = (ConnectivityManager) context
                .getSystemService(Context.CONNECTIVITY_SERVICE);
        subscriptionManager = SubscriptionManager.from(context);
        resolver = context.getContentResolver();
        isMultiSim = false;

        try {
            Method method = TelephonyManager.class.getDeclaredMethod("isMultiSimEnabled");
            isMultiSim = (boolean) method.invoke(telephonyManager);
            LogUtil.d(TAG, "isMultiSimEnabled:" + isMultiSim);
        } catch (NoSuchMethodException e) {
        } catch (IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }

    }

    @Override
    public String getImei() {
        String imei = "";
        try {
            Method getImei = TelephonyManager.class.getDeclaredMethod("getImei", Integer.TYPE);
            imei = (String) getImei.invoke(this.telephonyManager, 0);
        } catch (Exception e) {
            Log.e(TAG, "Reflection exception" + e.getMessage());
        }
        return imei;
    }


    @Override
    public boolean getDataEnabled(int subId) {
        boolean result = false;
        try {
            Method method = TelephonyManager.class.getDeclaredMethod("getDataEnabled", Integer.TYPE);
            result = (boolean) method.invoke(this.telephonyManager, subId);
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
        return result;
    }

    @Override
    public void setDataEnabled(int subId, boolean enable) {
        try {
            Method method = TelephonyManager.class
                    .getDeclaredMethod("setDataEnabled", Integer.TYPE, Boolean.TYPE);
            method.invoke(telephonyManager, subId, enable);
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean getDataRoamingEnabled(int subId) {
        boolean result = false;
        if (!isMultiSim) {
            result = Settings.Global.getInt(this.resolver, "data_roaming", 0) != 0;
        } else {
            try {
                Method method = TelephonyManager.class
                        .getDeclaredMethod("getIntWithSubId", ContentResolver.class, String.class,
                                Integer.TYPE);
                result = (boolean) method
                        .invoke(this.telephonyManager, this.resolver, "data_roaming", subId);
            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        }
        return result;
    }

    @Override
    public void setDataRoamingEnabled(int subId, boolean enable) {
        int roaming = enable ? 1 : 0;
        try {
            Settings.Global.putInt(this.resolver,
                    isMultiSim ? Settings.Global.DATA_ROAMING + subId : Settings.Global.DATA_ROAMING,
                    roaming);
        } catch (SecurityException e) {
            e.printStackTrace();
        }

        try {
            @SuppressLint("DiscouragedPrivateApi") Method method = SubscriptionManager.class
                    .getDeclaredMethod("setDataRoaming", Integer.TYPE, Integer.TYPE);
            method.invoke(this.subscriptionManager, roaming, subId);
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean setNonRoamingForOperator(String mccMnc) {
        //TODO not implemented
        return false;
    }

    @Override
    public boolean setOperatorBrandName(String name) {
        return telephonyManager.setOperatorBrandOverride(name);
    }

    @Override
    public int getSubId(int slotId) {
        try {
            Method method = SubscriptionManager.class.getDeclaredMethod("getSubId", Integer.TYPE);
            int[] subIdArray = (int[]) method.invoke(null, slotId);
            if (subIdArray != null && subIdArray.length > 0) {
                return subIdArray[0];
            }
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
        return -1;
    }

    @Override
    public String getImsi(int subId) {
        String imsi = null;
        try {
            Method method = TelephonyManager.class.getDeclaredMethod("getSubscriberId", Integer.TYPE);
            imsi = (String) method.invoke(this.telephonyManager, subId);
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
        return imsi;
    }

    @Override
    public String initIccid() {
        String iccId = telephonyManager.getSimSerialNumber();
            SubscriptionInfo info = subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(0);
            if (info != null) {
                // 在插入卡片的时候多数情况会不对
                iccId = info.getIccId();
            }
        if (!TextUtils.isEmpty(iccId) && !TextUtils.equals(iccId, SharePrefSetting.getCurrentIccId())) {
            SharePrefSetting.putCurrentIccId(iccId);
        }
        return iccId;
    }

    @Override
    public boolean isSimReady(int slotId) {
        int simState = 0;
        try {
            Method method = TelephonyManager.class.getDeclaredMethod("getSimState", Integer.TYPE);
            simState = (int) method.invoke(this.telephonyManager, slotId);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return simState == 5;
    }

    @Override
    public int getDefaultDataSubId() {
        int subId = -1;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            subId = SubscriptionManager.getDefaultDataSubscriptionId();
            Log.e(TAG, String.format("getDefaultDataSubId() returns %d", subId));
            return subId;
        }
        try {
            Method method = SubscriptionManager.class.getDeclaredMethod("getDefaultDataSubId");
            subId = (int) method.invoke(this.subscriptionManager);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return subId;
    }

    @Override
    public void setDefaultDataSubId(int subId) {
        if (subId < 0) {
            return;
        }
        try {
            Method method = SubscriptionManager.class
                    .getDeclaredMethod("setDefaultDataSubId", Integer.TYPE);
            method.invoke(this.subscriptionManager, subId);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void setDisplayName(int subId, String name) {
        try {
            @SuppressLint("SoonBlockedPrivateApi") Method method = SubscriptionManager.class
                    .getDeclaredMethod("setDisplayName", String.class, Integer.TYPE);
            method.invoke(this.subscriptionManager, name, subId);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public String getModel() {
        return Build.BRAND;
    }

    @Override
    public String getNetworkType() {
        String netType = "UNKNOWN";
        NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
        if (networkInfo != null && networkInfo.isConnected()) {
            if (networkInfo.getType() == ConnectivityManager.TYPE_WIFI) {
                netType = "WIFI";
            } else if (networkInfo.getType() == ConnectivityManager.TYPE_MOBILE) {
                int subtype = networkInfo.getSubtype();
                switch (subtype) {
                    case NETWORK_TYPE_GSM:
                    case NETWORK_TYPE_GPRS:
                    case NETWORK_TYPE_CDMA:
                    case NETWORK_TYPE_EDGE:
                    case NETWORK_TYPE_1xRTT:
                    case NETWORK_TYPE_IDEN:
                        netType = "2G";
                        break;
                    case NETWORK_TYPE_TD_SCDMA:
                    case NETWORK_TYPE_EVDO_A:
                    case TelephonyManager.NETWORK_TYPE_UMTS:
                    case TelephonyManager.NETWORK_TYPE_EVDO_0:
                    case TelephonyManager.NETWORK_TYPE_HSDPA:
                    case TelephonyManager.NETWORK_TYPE_HSUPA:
                    case TelephonyManager.NETWORK_TYPE_HSPA:
                    case TelephonyManager.NETWORK_TYPE_EVDO_B:
                    case TelephonyManager.NETWORK_TYPE_EHRPD:
                    case TelephonyManager.NETWORK_TYPE_HSPAP:
                        netType = "3G";
                        break;
                    case NETWORK_TYPE_IWLAN:
                    case NETWORK_TYPE_LTE:
                        netType = "4G";
                        break;
                    default:
                        String subtypeName = networkInfo.getSubtypeName();
                        if (subtypeName.equalsIgnoreCase("TD-SCDMA")
                                || subtypeName.equalsIgnoreCase("WCDMA")
                                || subtypeName.equalsIgnoreCase("CDMA2000")) {
                            netType = "3G";
                        }
                        break;
                }

            }
        }
        return netType;
    }

    @Override
    public boolean configureApn(String apn, String mccMnc) {
        int subId = getSubId(0);
        if (subId < 0) {
            return false;
        }
        if (TextUtils.isEmpty(mccMnc) || TextUtils.isEmpty(apn)) {
            return false;
        }
        int apnId = lookupApn(apn, mccMnc);
        if (apnId < 0) {
            apnId = insertApn(apn, mccMnc);
        }

        return apnId >= 0 && setPreferredApn(subId, apnId);
    }

    @Override
    public boolean clearApn() {
        int delete = resolver.delete(APN_URI, "name = ?", new String[]{APN_NAME});
        return delete >= 0;
    }

    private int lookupApn(String apn, String mccMnc) {
        Cursor cursor = resolver.query(APN_URI, null, "name = ? AND apn = ? AND numeric = ?",
                new String[]{APN_NAME, apn, mccMnc}, null);
        if (cursor == null) {
            return -1;
        } else {
            int id = -1;
            if (cursor.moveToFirst()) {
                id = cursor.getShort(cursor.getColumnIndex("_id"));
            }
            cursor.close();
            return id;
        }
    }

    private int insertApn(String apn, String mccMnc) {
        ContentValues values = new ContentValues();
        values.put("name", APN_NAME);
        values.put("apn", apn);
        values.put("numeric", mccMnc);
        values.put("mcc", mccMnc.substring(0, 3));
        values.put("mnc", mccMnc.substring(3));
        values.put("type", "default");
        Uri insert = resolver.insert(APN_URI, values);

        if (insert == null) {
            return -1;
        } else {
            Cursor cursor = resolver.query(insert, null, null, null, null);
            int id = -1;
            if (cursor.moveToFirst()) {
                id = cursor.getShort(cursor.getColumnIndex("_id"));
            }
            cursor.close();
            return id;
        }
    }

    private boolean setPreferredApn(int subId, int apnId) {
        ContentValues values = new ContentValues();
        values.put("apn_id", apnId);
        Uri uri = this.isMultiSim ? Uri.withAppendedPath(PREFERRED_APN_URI, "subId/" + subId)
                : PREFERRED_APN_URI;
        int update = resolver.update(uri, values, null, null);
        return update > 0;
    }
}

File: app/src/main/java/com/redteamobile/smart/agent/cellular/TelephonySettingImpl.kt
package com.redteamobile.smart.agent.cellular

import android.annotation.SuppressLint
import android.content.ContentResolver
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.net.ConnectivityManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.telephony.SubscriptionManager
import android.telephony.TelephonyManager
import android.telephony.TelephonyManager.*
import android.text.TextUtils
import android.util.Log
import com.redteamobile.smart.agent.util.LogUtil
import com.redteamobile.smart.agent.util.SharePrefSetting
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import kotlin.math.log


class TelephonySettingImpl() : TelephonySetting {

    private val TAG = "TelephonySettingImpl"
    private val APN_NAME = "Redtea Mobile"
    private val APN_URI = Uri.parse("content://telephony/carriers")
    private val PREFERRED_APN_URI = Uri.parse("content://telephony/carriers/preferapn")

    private var telephonyManager: TelephonyManager? = null
    private var subscriptionManager: SubscriptionManager? = null
    private var connectivityManager: ConnectivityManager? = null
    private var resolver: ContentResolver? = null
    private var isMultiSim: Boolean = false

    constructor(context: Context) : this() {
        telephonyManager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
        connectivityManager =
            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        subscriptionManager = SubscriptionManager.from(context)
        resolver = context.contentResolver
        var isMultiSim = false
        try {
            val method = TelephonyManager::class.java.getDeclaredMethod("isMultiSimEnabled")
            isMultiSim = method.invoke(
                context.getSystemService(Context.TELEPHONY_SERVICE)
            ) as Boolean
            LogUtil.d(TAG, String.format("isMultiSimEnabled: %b", isMultiSim))
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

        this.isMultiSim = isMultiSim
    }

    override fun getImei(): String {
        var getImei: Method? = null
        try {
            getImei = TelephonyManager::class.java.getDeclaredMethod(
                "getImei", Int::class.javaPrimitiveType
            )
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection exception", e)
        }

        var any: Any? = null
        try {
            any = getImei!!.invoke(telephonyManager, 0)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection exception", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection exception", e)
        }
        LogUtil.e(TAG, "imei:" + any.toString())
        return if (any != null && any is String) {
            any.toString()
        } else ""
    }

    override fun getDataEnabled(subId: Int): Boolean {
        var result = false
        try {
            val method = TelephonyManager::class.java.getDeclaredMethod(
                "getDataEnabled",
                Int::class.javaPrimitiveType
            )
            result = method.invoke(telephonyManager, subId) as Boolean
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }
        LogUtil.i(TAG, String.format("getDataEnabled(subId: %d) returns %b", subId, result))
        return result
    }

    override fun setDataEnabled(subId: Int, enable: Boolean) {
        LogUtil.i(TAG, String.format("setDataEnabled(subId: %d, enable: %b)", subId, enable))
        try {
            val method = TelephonyManager::class.java.getDeclaredMethod(
                "setDataEnabled", Int::class.javaPrimitiveType, Boolean::class.javaPrimitiveType
            )
            method.invoke(telephonyManager, subId, enable)
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

    }

    override fun getDataRoamingEnabled(subId: Int): Boolean {
        var result = false
        if (!isMultiSim) {
            result = Settings.Global.getInt(resolver, Settings.Global.DATA_ROAMING, 0) != 0
        } else {
            try {
                val method = TelephonyManager::class.java.getDeclaredMethod(
                    "getIntWithSubId",
                    ContentResolver::class.java,
                    String::class.java,
                    Int::class.javaPrimitiveType
                )
                result = method.invoke(
                    telephonyManager, resolver, Settings.Global.DATA_ROAMING, subId
                ) as Int != 0
            } catch (e: NoSuchMethodException) {
                LogUtil.e(TAG, "Reflection failed", e)
            } catch (e: IllegalAccessException) {
                LogUtil.e(TAG, "Reflection failed", e)
            } catch (e: InvocationTargetException) {
                LogUtil.e(TAG, "Reflection failed", e)
            }

        }
        LogUtil.i(TAG, String.format("getDataRoamingEnabled(subId: %d) returns %b", subId, result))
        return result
    }

    override fun setDataRoamingEnabled(subId: Int, enable: Boolean) {
        LogUtil.i(
            TAG, String.format(
                "setDataRoamingEnabled(subId: %d, enable: %b)", subId, enable
            )
        )
        val roaming = if (enable) 1 else 0
        try {
            Settings.Global.putInt(
                resolver, if (isMultiSim)
                    Settings.Global.DATA_ROAMING + subId
                else
                    Settings.Global.DATA_ROAMING, roaming
            )
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied", e)
        }

        try {
            val method = SubscriptionManager::class.java.getDeclaredMethod(
                "setDataRoaming", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType
            )
            method.invoke(subscriptionManager, roaming, subId)
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

    }

    override fun setNonRoamingForOperator(mccMnc: String): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun setOperatorBrandName(name: String): Boolean {
        var result = false
        try {
            result = telephonyManager!!.setOperatorBrandOverride(name)
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied with setOperatorBrandName()")
        }

        LogUtil.i(
            TAG, String.format(
                "setOperatorBrandName(name: %s) returns %b", name, result
            )
        )
        return result
    }

    @SuppressLint("MissingPermission")
    override fun getSubId(slotId: Int): Int {
        try {
            val clz = Class.forName("android.telephony.SubscriptionManager")
            val method = clz.getDeclaredMethod("getSubId", Int::class.javaPrimitiveType)
            val result = method.invoke(null, slotId)
                ?: return subscriptionManager?.getActiveSubscriptionInfoForSimSlotIndex(slotId)?.subscriptionId!!
            val subId: IntArray = result as IntArray
            if (subId != null && subId.isNotEmpty()) {
                return subId[0]
            }

        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

        return -1
    }

    override fun getImsi(subId: Int): String {
        var imsi: String? = null
        try {
            val method = TelephonyManager::class.java.getDeclaredMethod(
                "getSubscriberId",
                Int::class.javaPrimitiveType
            )
            imsi = method.invoke(telephonyManager, subId) as String?
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

        if (imsi == null) {
            imsi = ""
        }
        return imsi
    }

    @SuppressLint("MissingPermission")
    override fun initIccid() {
        var iccId: String
        iccId = telephonyManager?.simSerialNumber.toString()
        var info = subscriptionManager?.getActiveSubscriptionInfoForSimSlotIndex(0)
        Log.d(TAG, if (info == null) "NULL" else info.iccId)
        if (info != null) {
            iccId = info.iccId
        }
        if (iccId.isNotEmpty() && !TextUtils.equals(iccId, SharePrefSetting.getCurrentIccId())) {
            SharePrefSetting.putCurrentIccId(iccId.toUpperCase())
        }
    }

    override fun isSimReady(slotId: Int): Boolean {
        var simState = SIM_STATE_UNKNOWN
        try {
            val method = TelephonyManager::class.java.getDeclaredMethod(
                "getSimState",
                Int::class.javaPrimitiveType
            )
            simState = method.invoke(telephonyManager, slotId) as Int
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

        val result = simState == SIM_STATE_READY
        LogUtil.i(TAG, String.format("isSimReady(slotId: %d) returns %b", slotId, result))
        return result
    }

    override fun getDefaultDataSubId(): Int {
        var subId = -1
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            subId = SubscriptionManager.getDefaultDataSubscriptionId()
            LogUtil.i(TAG, String.format("getDefaultDataSubId() returns %d", subId))
            return subId
        }
        try {
            val method = SubscriptionManager::class.java.getDeclaredMethod("getDefaultDataSubId")
            subId = method.invoke(subscriptionManager) as Int
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

        LogUtil.i(TAG, String.format("getDefaultDataSubId() returns %d", subId))
        return subId
    }

    override fun setDefaultDataSubId(subId: Int) {
        LogUtil.i(TAG, String.format("setDefaultDataSubId(subId: %d)", subId))
        if (subId < 0) {
            return
        }
        try {
            val method = SubscriptionManager::class.java.getDeclaredMethod(
                "setDefaultDataSubId", Int::class.javaPrimitiveType
            )
            method.invoke(subscriptionManager, subId)
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

    }

    @SuppressLint("SoonBlockedPrivateApi")
    override fun setDisplayName(subId: Int, name: String) {
        LogUtil.i(TAG, String.format("setDisplayName(subId: %d, name: %s)", subId, name))
        try {
            val method = SubscriptionManager::class.java.getDeclaredMethod(
                "setDisplayName", String::class.java, Int::class.javaPrimitiveType
            )
            method.invoke(subscriptionManager, name, subId)
        } catch (e: NoSuchMethodException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: IllegalAccessException) {
            LogUtil.e(TAG, "Reflection failed", e)
        } catch (e: InvocationTargetException) {
            LogUtil.e(TAG, "Reflection failed", e)
        }

    }

    override fun getModel(): String {
        return Build.BRAND
    }

    override fun getNetworkType(): String {
        val info = connectivityManager?.activeNetworkInfo
        var netType = ""
        if (info != null && info.isConnected) {
            if (info.type == ConnectivityManager.TYPE_WIFI) {
                netType = "WIFI"
            } else if (info.type == ConnectivityManager.TYPE_MOBILE) {
                when (info.subtype) {
                    NETWORK_TYPE_GSM, NETWORK_TYPE_GPRS, NETWORK_TYPE_CDMA, NETWORK_TYPE_EDGE,
                    NETWORK_TYPE_1xRTT, NETWORK_TYPE_IDEN -> netType = "2G"

                    NETWORK_TYPE_TD_SCDMA, NETWORK_TYPE_EVDO_A, NETWORK_TYPE_UMTS,
                    NETWORK_TYPE_EVDO_0, NETWORK_TYPE_HSDPA, NETWORK_TYPE_HSUPA,
                    NETWORK_TYPE_HSPA, NETWORK_TYPE_EVDO_B, NETWORK_TYPE_EHRPD, NETWORK_TYPE_HSPAP -> netType =
                        "3G"

                    NETWORK_TYPE_IWLAN, NETWORK_TYPE_LTE -> netType = "4G"
                    else -> {
                        val subtypeName = info.subtypeName
                        netType = if (subtypeName.equals("TD-SCDMA", ignoreCase = true)
                            || subtypeName.equals("WCDMA", ignoreCase = true)
                            || subtypeName.equals("CDMA2000", ignoreCase = true)
                        ) {
                            "3G"
                        } else {
                            "UNKNOWN"
                        }
                    }
                }

            }
        } else {
            netType = "UNKNOWN"
        }
        return netType
    }

    override fun configureApn(apn: String, mccMnc: String): Boolean {
        val subId = getSubId(0)
        LogUtil.e(TAG, "subId:$subId mccMnc:$mccMnc")
        if (subId < 0) {
            return false
        }
        if (TextUtils.isEmpty(mccMnc) || TextUtils.isEmpty(apn)) {
            return false
        }
        var apnId = lookupApn(apn, mccMnc)
        if (apnId < 0) {
            apnId = insertApn(apn, mccMnc)
        }

        val result = apnId >= 0 && setPreferredApn(subId, apnId)
        LogUtil.i(
            TAG, String.format(
                "configureApn(apn: %s, mccMnc: %s, subId: %d) returns %b",
                apn, mccMnc, subId, result
            )
        )
        return result
    }

    override fun clearApn(): Boolean {
        var result = false
        try {
            result = resolver?.delete(APN_URI, "name = ?", arrayOf(APN_NAME))!! >= 0
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied", e)
        }

        LogUtil.i(TAG, String.format("clearApn() returns %b", result))
        return result
    }

    private fun lookupApn(apn: String, mccMnc: String): Int {
        LogUtil.i(TAG, "lookupApn()")

        var cursor: Cursor? = null
        try {
            cursor = resolver?.query(
                APN_URI, null, "name = ? AND apn = ? AND numeric = ?",
                arrayOf(APN_NAME, apn, mccMnc), null
            )
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied", e)
        }

        if (cursor == null) {
            return -1
        }
        var id = -1
        if (cursor.moveToFirst()) {
            id = cursor.getShort(cursor.getColumnIndex("_id")).toInt()
        }
        cursor.close()
        return id
    }

    private fun insertApn(apn: String, mccMnc: String): Int {
        LogUtil.i(TAG, "insertApn()")
        val values = ContentValues()
        values.put("name", APN_NAME)
        values.put("apn", apn)
        values.put("numeric", mccMnc)
        values.put("mcc", mccMnc.substring(0, 3))
        values.put("mnc", mccMnc.substring(3))
        values.put("type", "default")

        var newRow: Uri? = null
        try {
            newRow = resolver?.insert(APN_URI, values)
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied", e)
        }

        if (newRow == null) {
            return -1
        }

        var cursor: Cursor? = null
        try {
            cursor = resolver?.query(newRow, null, null, null, null)
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied", e)
        }

        if (cursor == null) {
            return -1
        }
        var id = -1
        if (cursor.moveToFirst()) {
            id = cursor.getShort(cursor.getColumnIndex("_id")).toInt()
        }
        cursor.close()
        return id
    }

    private fun setPreferredApn(subId: Int, apnId: Int): Boolean {
        LogUtil.i(TAG, "setPreferredApn()")

        val values = ContentValues()
        values.put("apn_id", apnId)
        val uri = if (isMultiSim)
            Uri.withAppendedPath(PREFERRED_APN_URI, "subId/$subId")
        else
            PREFERRED_APN_URI

        var result = false
        try {
            result = resolver?.update(uri, values, null, null)!! > 0
        } catch (e: SecurityException) {
            LogUtil.e(TAG, "Permission denied", e)
        }

        return result
    }

}
